design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+in order to take advantage of a shared connection pool and other properties if desired.
+
+`ClientOptions` will initially be as follows:
+```
+public interface ClientOptions {
+
+  ClientOptions setHost(String host);
+
+  ClientOptions setPort(int port);
+
+  ClientOptions setUseTls(boolean useTls);
+
+  ClientOptions setUseClientAuth(boolean useClientAuth);
+
+  ClientOptions setTrustStore(String trustStorePath);
+
+  ClientOptions setTrustStorePassword(String trustStorePassword);
+
+  ClientOptions setKeyStore(String keyStorePath);
+
+  ClientOptions setKeyStorePassword(String keyStorePassword);
+
+  ClientOptions setBasicAuthCredentials(String username, String password);
+
+  String getHost();
+
+  int getPort();
+
+  boolean isUseTls();
+
+  boolean isUseClientAuth();
+
+  boolean isUseBasicAuth();
+
+  String getTrustStore();
+
+  String getTrustStorePassword();
+
+  String getKeyStore();
+
+  String getKeyStorePassword();
+
+  String getBasicAuthUsername();
+
+  String getBasicAuthPassword();
+
+  ClientOptions copy();
+
+  static ClientOptions create() {
+    return new ClientOptionsImpl();
+  }
+}
+```
+
+We can always add additional configuration options later. We may also wish to expose the Vert.x `HttpClientOptions` for
+advanced users that wish to provide custom configs.
+
+### Transient queries -- Streaming
+
+The `Client` interface will provide the following methods for streaming the results of a transient (push or pull) query:
+```
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql, Map<String, Object> properties);
+```
+where `QueryResult` is as follows:
+```
+/**
+ * The result of a query (push or pull), streamed one row at time. Records may be consumed by either
+ * subscribing to the publisher or polling (blocking) for one record at a time. These two methods of
+ * consumption are mutually exclusive; only one method may be used (per QueryResult).
+ */
+public interface QueryResult extends Publisher<Row> {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  String queryID();
+
+  /**
+   * Block until a row becomes available.
+   *
+   * @return the row.
+   */
+  Row poll();
+
+  /**
+   * Block until a row becomes available or the timeout has elapsed.
+   *
+   * @param timeout amount of to wait for a row. Non-positive values are interpreted as no timeout.
+   * @param timeUnit unit for timeout param.
+   * @return the row, if available; else, null.
+   */
+  Row poll(long timeout, TimeUnit timeUnit);
```

Comment:
I would prefer it to return null in case of timeout. The timeout here is not an exceptional condition, e.g. it would be common for the user to call poll with a low timeout to keep the callers thread live (maybe it needs to poll multiple consumers? Aside: Personally I don't like the synchronous way, but many people do, and it's the way Kafka consumers are accessed, so users will likely be used to this).
````
while (true) {
  for (Consumer consumer: consumers) {
     Row row = consumer.poll(0);
     if (row != null) {
       doSomethingWithRow(row);
     }
  }
}
````
In the above example it's expected that row will be null, probably in most cases. It's not an exceptional condition. (Returning null (or some other return value representing no value) is pretty common in these kind of polling APIs).
If the method threw an exception it would make the above code somewhat more complex, also performance wise, creating and throwing exceptions is quite expensive (mainly due to stack trace population).

Reply:
It might be reasonable to treat a timeout as non exceptional. However, returning `null` seems like an anti-pattern to me. In doubt, the return type should be an `Optinal`?

> and it's the way Kafka consumers are accessed, so users will likely be used to this access style

The example is not how KafkaConsumer works: `poll()` returns a not-null `ConsumerRecords` object that implement `Iterable` -- thus, there would never be a `null` returned to the application.

Reply:
Btw: instead of using two parameters, it would be simpler to use a single `Duration`.

One more after thought: the parameter name "timeout" might sound like an "exceptional condition", thus, we could also pick a different name like `maxBlock` or similar? Just a random thought.

Reply:
> The example is not how KafkaConsumer works: poll() returns a not-null ConsumerRecords object that implement Iterable -- thus, there would never be a null returned to the application.

The sentence was referring to the Kafka consumer being accessed in a synchronous way and not throwing an exception.

Reply:
The Kafka consumer is always sync in `poll()` and never throws a `TimeoutException`. If the `poll()` wait time expires, it just returns a `ConsumerRecords` object with zero records in it.

Reply:
> However, returning null seems like an anti-pattern to me. In doubt, the return type should be an Optional?

Ah the old "null vs optional debate"! Personally, I'm not a massive fan of Optional, I think the hype around dealing with nulls, is.. well, overhyped.

Maybe I'm just old school, I learnt my Java before Optional existed and tbh, I never really recognised the problem it was fixing as an actually problem I had ever really seen that much or really cared about.

Also, for performance sensitive code, returning an Optional creates garbage unlike returning null. So for the more performance minded users this may be a turn-off.

Having said that, I wouldn't consider it the end of the world if we used Optional here.

Reply:
IMHO, it's always good to avoid `null` to pro-actively avoid NPEs. And I doubt that it will be an actual performance concern. But I am not religious about it, and if we want to use `null` so be it; I just have the impression that "modern Java" is moving to use `Optional` and it might be good to align to this development.

Reply:
One of the silly things about Optional imho:

Yes, it prevents NPE, but then if you call .get() on the optional and it doesn't have a value it throws NoSuchElementException!

So you're just trading the NPE for some other exception. Seems rather pointless to me!

It's only if you use Optional in a completely functional way. I.e never call get() and use the methods that take different actions based on whether the value is set or not, that it has value. But it's rarely used completely in that way in my experience.

Reply:
Well, I think the advantage is that the return type itself documents if it might be `null` or not. Hence, the overall contract should be that any non-optional return type is never `null`. This make the contract of nullable explicit; otherwise, you always need to check for `null` just to be sure (of course, it's documented in the JavaDocs, but a code level contract seems preferable to a "comment contract".

But as so often: it's subjective.

Reply:
>> Hence, the overall contract should be that any non-optional return type is never null.

Imho, unless these kinds of contracts are enforced by the compiler (as they are with Rust or Kotlin) they have little value imho. But we agree on one thing - it's subjective :)

Reply:
Agree. Compiler enforcement would be best! (Similar for "read only" objects... Java is just very bad at those things...)